### PR TITLE
Godot compiler plugin

### LINKED ISSUE #72: Implement kotlin compiler plugins
Implement kotlin compiler plugins like here:
https://github.com/utopia-rise/godot-kotlin/tree/master/tools/kotlin-compiler-native-plugin
and here:
https://github.com/utopia-rise/godot-kotlin/tree/master/tools/kotlin-compiler-plugin
using the [MpApt](https://github.com/Foso/MpApt) project.

### PR DESCRIPTION
**Pull request by chippmann**
**Important:** This is not ready to be merged until this [PR](https://github.com/Foso/MpApt/pull/15) in the MpApt project is merged and published. Still it can be reviewed until then as once the PR is merged it's just a version change on our side.

This implements the godot compiler plugin and closes #72.

### CODE REVIEW

File: settings.gradle.kts
Comment (raniejade) [inline settings.gradle.kts]: You can collapse these lines into a single `subdir`.

```suggestion
subdir("entry-generation") {
    include("godot-compiler-native-plugin")
    include("godot-compiler-plugin")
    include("godot-compiler-plugin-common")
}
```

File: plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt
Comment (raniejade) [inline plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt:43]: check `task` if it is a instance of `KotlinNativeCompile`?

File: entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt
Comment (raniejade) [inline entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt]: use `checkNotNull(...) { "error message here" }`, `!!` just throws an NPE which is not really helpful for debugging.

File: entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt
Comment (raniejade) [inline entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt]: what does `<name>` mean here? If it is used as a help text, maybe use a more descriptive text? i.e since this is for gdns dir, maybe `<path-to-gdns-dir>`.

File: entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt
Comment (raniejade) [inline entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt]: same as above.

File: entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt
Comment (raniejade) [inline entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt]: not a fan of `compilerplugin` subpackage. How about `godot.kotlin.compiler`?

**Review by raniejade — CHANGES_REQUESTED** [review]
A few comments, but overall LGTM!

File: build.gradle.kts
Comment (raniejade) [inline build.gradle.kts]: Something for later, but we can make this so that we only need to manage a single version. The idea here is that `CompilerPluginConst` will load the version value from a `build.properties` file included in the jar. The properties file is then processed by gradle via the standard task `processResources`. Relevant sources from my binding:

[1] https://github.com/raniejade/godot-kotlin/blob/master/godot-kotlin-gradle-plugin/src/main/resources/build.properties
[2] https://github.com/raniejade/godot-kotlin/blob/master/godot-kotlin-gradle-plugin/build.gradle.kts#L55
[3] https://github.com/raniejade/godot-kotlin/blob/master/godot-kotlin-gradle-plugin/src/main/kotlin/godot/gradle/buildProperties.kt#L6

File: settings.gradle.kts
Comment (chippmann) [reply settings.gradle.kts]: :+1:

File: plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt
Comment (chippmann) [reply plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt:43]: :+1:

File: entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt
Comment (chippmann) [reply entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt]: True! Good idea

File: entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt
Comment (chippmann) [reply entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt]: true! I changed it.

File: plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt
Comment (chippmann) [reply plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt:43]: But shouldn't it be `KotlinNativeCompilation`?

File: entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt
Comment (chippmann) [reply entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compilerplugin/CommonComponentRegistrar.kt]: agreed

File: build.gradle.kts
Comment (chippmann) [reply build.gradle.kts]: I looked for an elegant solution but this didn't came to my mind... I really like it. As i have to wait for MpApt to merge. I'll include this in this PR as well. This is billiant!

File: plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt
Comment (raniejade) [reply plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt:43]: Yeah, I keep forgetting the task class name :joy:

Comment (chippmann) [reply plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt:43]: ah no you were right. it expects a clas inherited from abstractCompile

Comment (chippmann) [reply plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt:43]: as discussed on discord: sadly not possible

File: entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compiler/plugin/CommonComponentRegistrar.kt
Comment (raniejade) [inline entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compiler/plugin/CommonComponentRegistrar.kt:12]: wait, is this required? As I can see it's a duplicate of NativeComponentRegistrar.

Comment (chippmann) [reply entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compiler/plugin/CommonComponentRegistrar.kt:12]: Yes it's basically the same. The only real difference is what components are registered. We basically only need the Native one but for compiler plugins you need to specify both if you want to use the native one.
this method is required: https://github.com/utopia-rise/godot-kotlin/blob/3ddfee05b972567090ca2b7f3a028aff6607c88d/plugins/godot-gradle-plugin/src/main/kotlin/godot/gradle/subplugin/GodotSubPlugin.kt#L45

Comment (raniejade) [reply entry-generation/godot-compiler-plugin/src/main/kotlin/godot/compiler/plugin/CommonComponentRegistrar.kt:12]: Ahhh, last question are those components required to make mpapt work?

**Review by raniejade — APPROVED** [review]

**Comment by chippmann** [comment]
Sorry saw this comment after the merge...
In our case no. But if we need the common one then yes.

**Pull request merged**
